jspwiki-main/src/main/java/org/apache/wiki/auth/user/XMLUserDatabase.java

```diff
@@ -91,27 +95,59 @@ public class XMLUserDatabase extends AbstractUserDatabase {
     private Document            c_dom;
     private File                c_file;
 
+    /**
+     * A lock used to ensure thread safety when accessing shared resources.
+     * This lock provides more flexibility and capabilities than the intrinsic locking mechanism,
+     * such as the ability to attempt to acquire a lock with a timeout, or to interrupt a thread
+     * waiting to acquire a lock.
+     *
+     * @see java.util.concurrent.locks.ReentrantLock
+     */
+    private final ReentrantLock lock;
+
+    public XMLUserDatabase() {
+        lock = new ReentrantLock();
+    }
+
     /** {@inheritDoc} */
-    @Override
-    public synchronized void deleteByLoginName( final String loginName ) throws WikiSecurityException {
-        if( c_dom == null ) {
-            throw new WikiSecurityException( "FATAL: database does not exist" );
-        }
+    public void deleteByLoginName(final String loginName) throws WikiSecurityException {
+        final AtomicBoolean userDeleted = new AtomicBoolean(false);
+        final AtomicReference<WikiSecurityException> exceptionRef = new AtomicReference<>();
 
-        final NodeList users = c_dom.getDocumentElement().getElementsByTagName( USER_TAG );
-        for( int i = 0; i < users.getLength(); i++ ) {
-            final Element user = ( Element )users.item( i );
-            if( user.getAttribute( LOGIN_NAME ).equals( loginName ) ) {
-                c_dom.getDocumentElement().removeChild( user );
+        Synchronizer.synchronize(lock, () -> {
+            try {
+                if (c_dom == null) {
+                    throw new WikiSecurityException("FATAL: database does not exist");
+                }
+
+                final NodeList users = c_dom.getDocumentElement().getElementsByTagName(USER_TAG);
+                for (int i = 0; i < users.getLength(); i++) {
+                    final Element user = (Element) users.item(i);
+                    if (user.getAttribute(LOGIN_NAME).equals(loginName)) {
+                        c_dom.getDocumentElement().removeChild(user);
 
-                // Commit to disk
-                saveDOM();
-                return;
+                        // Commit to disk
+                        saveDOM();
+                        userDeleted.set(true);
+                        return;
+                    }
+                }
+            } catch (WikiSecurityException e) {
+                exceptionRef.set(e);
```

Comment:
the eternal problem with lambdas, no checked exceptions... great way to deal with it!